strings/manacher.nim

```diff
@@ -0,0 +1,88 @@
+# Manacher's algorithm
+#[
+Determine the longest palindrome in a string in linear time in its length
+with Manacher's algorithm
+Inspired from:
+https://github.com/jilljenn/tryalgo/blob/master/tryalgo/manacher.py
+https://en.wikipedia.org/wiki/Longest_palindromic_substring
+]#
+import std/[strutils, sequtils, sets]
+{.push raises: [].}
+
+runnableExamples:
+  let example1 = "cabbbab"
+  doAssert(manacherString(example1) == "abbba")
+  doAssert(manacherIndex(example1) == 1 .. 5)
+  doAssert(manacherLength(example1) == 5)
+
+func manacherIndex*(s: string): HSlice[int, int] {.raises: [ValueError].} =
+    #[Longest palindrome in a string by Manacher
+      :param s: string, lowercase ascii, no whitespace
+      :returns: indexes i,j such that s[i:j] is the longest palindrome in s
+      :complexity: O(len(s))
```

Comment:
> O(len(s)) is actually not a proper way to describe the time complexity of an algorithm lol,

Why? Where do you get that from?
Single letter variable in time complexities are no better than single letter variable in code. That is even worse since there is no context to infer to what they refer to.

Sure thing, len(s) is not terrible, since I defined s after giving the complexity. RAM is almost never the limiting factor unless working in constrained environments like GPU and embedded architectures, so I naturally omit the time description from the complexity.

Reply:
For example bubble sort, do you say it's O(n^2) or O((len(s) - 1) * (len(s) - i- 1))? I mean sure, len(...) gives a clearer context, however for time and space complexity, these things are ignored. The time complexity of any let's say O(n) algorithm is O(kn), where k is the constant. Yet the K here is ignored. No matter how big the k is, the complexity is still O(n). I said it's better to write O(n) instead of O(len(s)) is because following the conventions is y ignorable lol. I recommend you to say "O(n), where n = len(s)" but sure, if you just write O(len(s)), it's succinct and recognizable as well! I think both way works haha!

Reply:
The complexity would not be the same if $n$ was the number of bits of the input string. If $n$ is always bound by a constant, the algorithm would be O(1). If we were doing multiple queries to the program with the same algorithm, maybe the complexity could be amortized. Sure. I know what is an asymptotic bound.

The thing is, complexity bounds tell a story. Sometimes some constant factor and it is actually important to mention it.
Since the len() might confuse people (if some people think about the bit complexity), I will put n and comment n after.

Reply:
Solved.